### ISSUE TITLE
Avoid breaking changes when one hal is upgraded

### ISSUE THREAD

**Issue by TeXitoi** [stm32-rs/stm32-usbd#14]
An idea for solving this problem. I had this idea reading https://github.com/atsamd-rs/atsamd/issues/11#issuecomment-533849284

A trait (or a set of trait) is present in a dedicated crate that expose the low level stm32 usb api of the hardware. Each hal provide a simple constructor that create a struct that implement this trait. Then, the stm32-usbd provide a constructor that take a type implementing the trait and do the hard stuffs.

Breaking changes in the hal is then transparent to stm32-usbd, it's only coupled with the crate describing the trait. This trait, as a trivial proxy of the hardware, should be stable (as the `-sys` crates).

I have no idea of the implementation of this crate, so I may tell something not doable, but in case it's working, I share the idea.

**Comment by Disasm** [comment]
Thanks for pointing out!
I have a pretty similar idea with this "trait" crate, but it's not clear how to express peripheral features. For example, old USBFS peripherals in STM32s has multiple boolean and u32 features (expressed as feature flags in this crate). They can be implemented as associated constants, but what if I face with a new feature later, which is not present in the trait? Additionally, traits are not mutual exclusive, but peripheral types are. This can cause some problems. I believe this is doable, but I need to figure out the exact way.

**Comment by twitchyliquid64** [comment]
Make a new trait for each new version of the API?

EG: What embedded-hal did for digital pins (the V1 and V2 traits, with a conversion trait implemented on V2)

**Comment by Disasm** [comment]
I came up with the following idea.
`stm32-usbd` provides a trait for peripheral, that provides an "enable and reset" method and pointers to the peripheral registers and endpoint memory. Each stm32 hal contains a struct with `USB` and `PINS` and implements this trait for the struct.
Turns out that with this construction endpoint memory access type is extremely difficult to pass around, so it's better to leave it as a crate feature.
@mvirkkunen, any thoughts?
